lib/ansible/modules/cloud/vmware/vmware_vm_vm_drs_rule.py

```diff
@@ -0,0 +1,362 @@
+            vms_list: List of VM names
+
+        Returns: List of VM managed objects
+
+        """
+        vm_obj_list = []
+        if vms_list is None:
+            vms_list = self.vm_list
+
+        for vm_name in vms_list:
+            vm_obj = find_vm_by_id(content=self.content, vm_id=vm_name,
+                                   vm_id_type='vm_name', cluster=self.cluster_obj)
+            if vm_obj is None:
+                self.module.fail_json(msg="Failed to find the virtual machine %s "
+                                          "in given cluster %s" % (vm_name,
+                                                                   self.cluster_name))
+            vm_obj_list.append(vm_obj)
+        return vm_obj_list
+
+    def get_rule_key_by_name(self, cluster_obj=None, rule_name=None):
+        """
+        Function to get a specific DRS rule key by name
+        Args:
+            rule_name: Name of rule
+            cluster_obj: Cluster managed object
+
+        Returns: Rule Object if found or None
+
+        """
+        if cluster_obj is None:
+            cluster_obj = self.cluster_obj
+
+        if rule_name:
+            rules_list = [rule for rule in cluster_obj.configuration.rule if rule.name == rule_name]
+            if rules_list:
+                return rules_list[0]
+        # No rule found
+        return None
+
+    @staticmethod
+    def normalize_rule_spec(rule_obj=None):
+        """
+        Function to return human readable rule spec
+        Args:
+            rule_obj: Rule managed object
+
+        Returns: Dictionary with Rule info
+
+        """
+        if rule_obj is None:
+            return {}
+        return dict(rule_key=rule_obj.key,
+                    rule_enabled=rule_obj.enabled,
+                    rule_name=rule_obj.name,
+                    rule_mandatory=rule_obj.mandatory,
+                    rule_uuid=rule_obj.ruleUuid,
+                    rule_vms=[vm.name for vm in rule_obj.vm],
+                    )
+
+    # Create
+    def create(self):
+        """
+        Function to create a DRS rule if rule does not exists
+        """
+        rule_obj = self.get_rule_key_by_name(rule_name=self.rule_name)
+        if rule_obj is not None:
+            # Rule already exists, remove and create again
+            # Cluster does not allow editing existing rule
+            existing_rule = self.normalize_rule_spec(rule_obj=rule_obj)
+            if ((sorted(existing_rule['rule_vms']) == sorted(self.vm_list)) or
+                    (existing_rule['rule_enabled'] != self.enabled) or
+                    (existing_rule['rule_mandatory'] != self.mandatory)):
+                # Delete existing rule as we cannot edit it
+                changed, result = self.delete(rule_name=self.rule_name)
+                if not changed:
+                    self.module.fail_json(msg="Failed to delete while updating rule %s due to %s" % (self.rule_name, result))
+        changed, result = self.create_rule_spec()
```

Comment:
@Akasurde as I understand it, you are comparing the existing rule with the user parameters, if the rule doesn't match → you delete the current rule and, re-create it with the user parameters.
If that so:

* `sorted(existing_rule['rule_vms']) == sorted(self.vm_list)` → `sorted(existing_rule['rule_vms']) != sorted(self.vm_list)`
* if the rule match the user parameter you shouldn't try to re-create it →
```
+            if ((sorted(existing_rule['rule_vms']) == sorted(self.vm_list)) and
+                    (existing_rule['rule_enabled'] == self.enabled) or
+                    (existing_rule['rule_mandatory'] == self.mandatory)):
+                self.module.exit_json(changed=False, rule_key=rule_key)
+            # Delete existing rule as we cannot edit it
+            changed, result = self.delete(rule_name=self.rule_name)
+            if not changed:
+                self.module.fail_json(msg="Failed to delete while updating rule %s due to %s" % (self.rule_name, result))
+        changed, result = self.create_rule_spec()
```
* Is there a reason you don't compare the affinity/anti-affinity state of the rule ?